File: kube/src/main/java/com/jacksonrakena/infrastructure/apps/persistence/Galahad.kt
package com.jacksonrakena.infrastructure.apps.persistence

import com.jacksonrakena.infrastructure.util.applyCommonConfiguration
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.ContainerPort
import org.cdk8s.plus28.ContainerProps
import org.cdk8s.plus28.Deployment
import org.cdk8s.plus28.DeploymentProps
import org.cdk8s.plus28.DeploymentStrategy
import org.cdk8s.plus28.EnvFrom
import org.cdk8s.plus28.ISecret
import org.cdk8s.plus28.PersistentVolumeClaim
import org.cdk8s.plus28.Protocol
import org.cdk8s.plus28.Service
import org.cdk8s.plus28.ServicePort
import org.cdk8s.plus28.ServiceProps
import org.cdk8s.plus28.ServiceType
import org.cdk8s.plus28.Volume
import org.cdk8s.plus28.VolumeMount
import software.constructs.Construct

class Galahad(
    scope: Construct,
    id: String,
    vaultSecret: ISecret,
    postgresSecret: ISecret,
    volumeClaim: PersistentVolumeClaim,
    props: ChartProps? = null,
) : Chart(scope, id, props) {
    val vol = Volume.fromPersistentVolumeClaim(this, "galahad-pvc-mount", volumeClaim)
    val deployment = Deployment(
        this,
        "galahad-deployment",
        DeploymentProps.builder()
            .replicas(1)
            .strategy(DeploymentStrategy.recreate())
            .volumes(listOf(vol))
            .containers(
                listOf(
                    ContainerProps.builder()
                        .name("vaultwarden")
                        .image("vaultwarden/server")
                        .applyCommonConfiguration()
                        .ports(
                            listOf(
                                ContainerPort.builder().number(80).build()
                            )
                        )
                        .envFrom(
                            listOf(
                                EnvFrom(null, null, vaultSecret)
                            )
                        )
                        .volumeMounts(
                            listOf(
                                VolumeMount.builder()
                                    .volume(vol)
                                    .readOnly(false)
                                    .path("/data")
                                    .subPath("vaultwarden-data")
                                    .build()
                            )
                        )
                        .build(),
                    ContainerProps.builder()
                        .name("postgres")
                        .image("postgres:15")
                        .applyCommonConfiguration()
                        .ports(
                            listOf(
                                ContainerPort.builder().number(5432).build()
                            )
                        )
                        .envVariables(
                            mapOf(
                                "POSTGRES_DB" to postgresSecret.envValue("db"),
                                "POSTGRES_USER" to postgresSecret.envValue("username"),
                                "POSTGRES_PASSWORD" to postgresSecret.envValue("password")
                            )
                        )
                        .volumeMounts(
                            listOf(
                                VolumeMount.builder()
                                    .volume(vol)
                                    .readOnly(false)
                                    .path("/var/lib/postgresql/data")
                                    .subPath("pg-data")
                                    .build()
                            )
                        )
                        .build()
                )
            )
            .build()
    )

    val postgresService = Service(
        this,
        "postgres-service",
        ServiceProps.builder()
            .selector(deployment)
            .ports(
                listOf(
                    ServicePort.builder().name("sql").protocol(Protocol.TCP).port(5432).targetPort(5432).build()
                )
            )
            .type(ServiceType.CLUSTER_IP)
            .build()
    )

    val vaultService = Service(
        this,
        "vaultwarden-service",
        ServiceProps.builder()
            .selector(deployment)
            .ports(
                listOf(
                    ServicePort.builder().name("web").port(80).targetPort(80).build()
                )
            )
            .type(ServiceType.CLUSTER_IP)
            .build()
    )
}

File: src/apps/persistence/galahad.ts
import { Construct } from "constructs";
import { Chart, ChartProps } from "cdk8s";
import * as kplus from "cdk8s-plus-28";
import { applyCommonConfiguration } from "../../util/container-utils";

export class Galahad extends Chart {
  public readonly postgresService: kplus.Service;
  public readonly vaultService: kplus.Service;

  constructor(
    scope: Construct,
    id: string,
    vaultSecret: kplus.ISecret,
    postgresSecret: kplus.ISecret,
    volumeClaim: kplus.PersistentVolumeClaim,
    props?: ChartProps
  ) {
    super(scope, id, props);

    const vol = kplus.Volume.fromPersistentVolumeClaim(
      this,
      "galahad-pvc-mount",
      volumeClaim
    );

    const deployment = new kplus.Deployment(this, "galahad-deployment", {
      replicas: 1,
      strategy: kplus.DeploymentStrategy.recreate(),
      volumes: [vol],
      containers: [
        applyCommonConfiguration({
          name: "vaultwarden",
          image: "docker.io/vaultwarden/server",
          ports: [{ number: 80 }],
          envFrom: [new kplus.EnvFrom(undefined, undefined, vaultSecret)],
          volumeMounts: [
            {
              volume: vol,
              path: "/data",
              subPath: "vaultwarden-data",
              readOnly: false,
            },
          ],
        }),
        applyCommonConfiguration({
          name: "postgres",
          image: "docker.io/library/postgres:15",
          ports: [{ number: 5432 }],
          envVariables: {
            POSTGRES_DB: postgresSecret.envValue("db"),
            POSTGRES_USER: postgresSecret.envValue("username"),
            POSTGRES_PASSWORD: postgresSecret.envValue("password"),
          },
          volumeMounts: [
            {
              volume: vol,
              path: "/var/lib/postgresql/data",
              subPath: "pg-data",
              readOnly: false,
            },
          ],
        }),
      ],
    });

    this.postgresService = new kplus.Service(this, "postgres-service", {
      selector: deployment,
      ports: [
        {
          name: "sql",
          protocol: kplus.Protocol.TCP,
          port: 5432,
          targetPort: 5432,
        },
      ],
      type: kplus.ServiceType.CLUSTER_IP,
    });

    this.vaultService = new kplus.Service(this, "vaultwarden-service", {
      selector: deployment,
      ports: [{ name: "web", port: 80, targetPort: 80 }],
      type: kplus.ServiceType.CLUSTER_IP,
    });
  }
}
